src/core/vanilla.ts

```diff
@@ -449,6 +449,10 @@ const writeAtomState = <Value, Update>(
       ((a: AnyWritableAtom, v: unknown) => {
         const isPendingPromisesExpired = !pendingPromises.length
         if (a === atom) {
+          if (!hasInitialValue(a)) {
```

Comment:
Can we enforce this with types aswell?

Reply:
Oh, I wish I could. The point is `a === atom`.
I think it's impossible, but I'm not sure. There might be a way. You can try as a type puzzle.

---
As a side note, I've been thinking, the jotai api is not super type friendly. It's still JS oriented. Especially function overload is tricky. See what they do in https://github.com/gaku-sei/re-jotai